src/iterators/bfs.jl

```diff
@@ -74,36 +77,24 @@ end
 Iterator to visit vertices in a graph using breadth-first search.
 """
 function Base.iterate(t::BFSIterator, state::BFSVertexIteratorState)
-    graph, visited, queue = t.graph, state.visited, state.queue
-    while !isempty(queue)
-        if state.n_visited == nv(graph)
-            return nothing
-        end
-        # we visit the first node in the queue
-        node_start = first(queue)
-        if !visited[node_start]
-            visited[node_start] = true
-            state.n_visited += 1
-            return (node_start, state)
-        end
-        # which means we arrive here when the first node was visited.
-        neigh = outneighbors(graph, node_start)
-        if state.neighbor_idx <= length(neigh)
-            node = neigh[state.neighbor_idx]
-            # we update the idx of the neighbor we will visit,
-            # if it is already visited, we repeat
-            state.neighbor_idx += 1
-            if !visited[node]
-                push!(queue, node)
-                state.visited[node] = true
-                state.n_visited += 1
-                return (node, state)
+    # we fill nodes in this level
+    if state.node_idx == length(state.curr_level)
+        state.n_visited += length(state.curr_level)
+        state.n_visited == nv(t.graph) && return nothing
+        @inbounds for node in state.curr_level
+            for adj_node in outneighbors(t.graph, node)
+                if !state.visited[adj_node]
+                    push!(state.next_level, adj_node)
+                    state.visited[adj_node] = true
+                end
             end
-        else
-            # when the first node and its neighbors are visited
-            # we remove the first node of the queue
-            popfirst!(queue)
-            state.neighbor_idx = 1
         end
+        length(state.next_level) == 0 && return nothing
+        state.curr_level, state.next_level = state.next_level, empty!(state.curr_level)
+        state.node_idx = 0
     end
+    # we visit all nodes in this level
+    state.node_idx += 1
+    @inbounds node = state.curr_level[state.node_idx]
```

Comment:
same remark on `@inbounds`, better to remove it

Reply:
how can we be sure that `state.node_ix in eachindex(state.curr_level)`?

Reply:
because it is sure that `state.node_ix < length(state.curr_level)` since we check that at the start of the function